_posts/2022-09-XX-rate-limiting-web-applications.markdown

```diff
@@ -0,0 +1,465 @@
+
+Too strict limits will annoy your users. Remember, you don't want to police the number of requests.
+You want fair usage of resources. You don't call the police when two toddlers fight over a toy.
+If they both need the toy, maybe it's fine to have multiple toys, so they don't have to fight over it.
+
+### Will you allow bursts or not?
+
+Depending on your application and endpoint, having one rate limit in place will be enough.
+For example, a global rate limit of 600 requests per minute may be perfect for every endpoint in your application.
+
+However, sometimes you may want to allow bursts.
+For example, when your mobile app starts, it performs some initial requests in rapid succession to get the latest data from your API, and after that it slows down.
+
+To handle these bursts, you may want to implement a "laddering" approach, and have multiple different limits in place:
+
+| Limit      | Operation A | Operation B | Operation C |
+|------------|-------------|-------------|-------------|
+| Per second | 10          | 10          | 100         |
+| Per minute | 60          | 60          | 500         |
+| Per hour   | 3600        | 600         | 500         |
+
+In the above table, a client could make 10 requests per second to *Operation A*.
+10 per second would normally translate to 36000 request per hour, but maybe at the hourly level, only 3600 is a better number.
+
+Again, measure, and don't prematurely add laddering. There's a good chance a single limit for all endpoints in your application may be sufficient.
+
+### What will be the partition/identifier?
+
+Previously, we used a *user identifier + action/operation identifier* to impose rate limits.
+There are many other request properties you can use to partition your requests:
+
+* Partition per endpoint
+* Partition per IP address
+  * Keep in mind users may be sharing an IP address, e.g. when behind a NAT/CGNAT/proxy
+* Partition per user
+  * Keep in mind you may have anonymous users, how will you distinguish those?
+* Partition per session
+  * What if your user starts a new session for every request?
+  * What if your user makes use of multiple devices with separate sessions?
+* Partition per browser
+* Partition per header (e.g. `X-Api-Token`), ...
+
+Also here, "it depends" on your application. A global rate limit per IP address may work for your application.
+More complex applications may need a combination of these, e.g. per-endpoint rate limiting combined with the current user.
+
+### Decide on exceptions
+
+Should rate limiting apply to all requests? Well... yes! We already discussed all endpoints in your application should be rate limited.
+
+A better question would be whether the same limits should apply for all types of users. As usual, the answer to this question will depend on your application.
+There is no silver bullet, but here are some examples to think about.
+
+Good candidates to have different rate limits in place:
```

Comment:
ALignment with your business model is something to emphasize (not perse as exception).

Premium customers are allowed to get more requests per second compared to free users

Reply:
Or premium does not get rate limited (on paper, always have a rate limit, even if it's ridiculously high)